backend/src/main/java/de/neuefische/backend/service/WatchHistoryService.java

```diff
@@ -58,4 +63,86 @@ public void removeFromWatchHistory(String imdbId){
                 .collect(Collectors.toList());
     }
 
+    public List<Episode> getWatchHistoryEpisodes(String imdbID, String season){
+        if(!movieAndSeriesRepo.existsById(imdbID)) {
+            return List.of();
+        }
+        return movieAndSeriesRepo.findByImdbID(imdbID)
+                .getWatchedEpisodes()
+                .stream()
+                .filter((episode) -> String.valueOf(episode.getSeason_number()).equals(season))
+                .collect(Collectors.toList());
+    }
+
+    public List<Episode> getWatchHistoryAllEpisodes(String imdbID){
+        if(!movieAndSeriesRepo.existsById(imdbID)) {
+            return List.of();
+        }
+        return movieAndSeriesRepo.findByImdbID(imdbID).getWatchedEpisodes();
+    }
+
+    public float getWatchHistorySeasonProgress(String imdbId, String tmdbId, String season){
+        if (getWatchHistoryEpisodes(imdbId, season) == null){
+            return 0;
+        }
+        float watched = getWatchHistoryEpisodes(imdbId, season).size();
+        float total = tmdbApiService.getEpisodes(tmdbId, season).getEpisodes().size();
+        return watched/total * 100;
+    }
+
+    public float getWatchHistoryTotalProgress(String imdbId, String tmdbId){
+        List<Episode> list = getWatchHistoryAllEpisodes(imdbId);
+        if (list == null){
+            return 0;
+        }
+        float watched = list.size();
+        float total = tmdbApiService.getDetails(tmdbId, "series").getNumber_of_episodes();
+        if (total == watched){
+            MovieAndSeries movieAndSeries = movieAndSeriesRepo.findByImdbID(imdbId);
+            movieAndSeries.setWatchHistory(true);
+            movieAndSeries.setWatching(false);
+            movieAndSeriesRepo.save(movieAndSeries);
+        }
+        return watched/total * 100;
+    }
+
+    public Episode addToWatchHistoryEpisodes(Episode episode){
+        MovieAndSeries movieAndSeries = movieAndSeriesRepo.findByImdbID(episode.getImdbId());
+        List<Episode> list = movieAndSeries.getWatchedEpisodes();
```

Comment:
for better code readability I would rename `list`to something more specific.

```suggestion
        List<Episode> watchedEpisodes = movieAndSeries.getWatchedEpisodes();
```